### PR TITLE
Xamarin.Mac Extension Support

### PR DESCRIPTION
**Pull request by chamons**
NO_MERGE until QA gives thumbs up.

This is enough work to get Today / Finder / Share extensions working. It can mostly be divided into a few buckets:

- Msbuild, must of it copy pasted from iOS with changes
- Runtime
- Bindings (for the Extension specific APIs)
- mmp

I plan on rebasing this into a single commit before landing, but I'd like feedback first.

There are auto tests in maccore (also PR) and a manual test script here https://docs.google.com/document/d/1o5A6cWssvm9MjDSZa9dw4YvOgg_OjdaFAc60_DN-gy4/edit#heading=h.cxiugjmzkun7

### CODE REVIEW

File: msbuild/Xamarin.Mac.Tasks/Tasks/Mmp.cs [hunk 1]
```diff
@@ -163,8 +165,10 @@ protected override string GenerateCommandLineCommands ()
 					args.AddQuoted ("/assembly:" + Path.GetFullPath (asm));
 			}
 
-			if (!string.IsNullOrEmpty (ApplicationAssembly))
-				args.AddQuoted (Path.GetFullPath (ApplicationAssembly));
+			if (!string.IsNullOrEmpty(ApplicationAssembly))
+			{
```
Comment (mandel-macaque) [inline msbuild/Xamarin.Mac.Tasks/Tasks/Mmp.cs]: Code style. You forgot the space between IsNullOrEmpty and the (

File: msbuild/Xamarin.Mac.Tasks/Tasks/Mmp.cs [hunk 2]
```diff
@@ -163,8 +165,10 @@ protected override string GenerateCommandLineCommands ()
...
+				args.AddQuoted(Path.GetFullPath(ApplicationAssembly + (IsAppExtension ? ".dll" : ".exe")));
+			}
```
Comment (mandel-macaque) [inline msbuild/Xamarin.Mac.Tasks/Tasks/Mmp.cs]: Same, style issue.

File: msbuild/Xamarin.Mac.Tasks/Xamarin.Mac.AppExtension.CSharp.targets [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+			Condition="Exists('$(MSBuildThisFileDirectory)$(MSBuildThisFileName).Before.targets')"/>
+
+	<PropertyGroup>
+		<!-- Version/fx properties -->
+		<TargetFrameworkVersion Condition="'$(TargetFrameworkIdentifier)' == '' And '$(TargetFrameworkVersion)' ==''">v4.5</TargetFrameworkVersion>
+		<!-- work around a bug in the Mono Microsoft.CSharp.Targets that defaults the compiler to gmcs -->
+		<!-- this is fixed in mono/master (2014-03-05), but we'll keep this workaround here for a while -->
+		<CscToolPath Condition="'$(OS)' == 'Unix'">/Library/Frameworks/Mono.framework/Commands</CscToolPath>
+		<CscToolExe Condition="'$(OS)' == 'Unix'">mcs</CscToolExe>
+
```
Comment (mandel-macaque) [inline msbuild/Xamarin.Mac.Tasks/Xamarin.Mac.AppExtension.CSharp.targets]: Small question, is this going to be used by VS? I'm wondering about when these two conditions are not met.

File: src/notificationcenter.cs [hunk 4]
```diff
@@ -47,6 +64,107 @@ interface UIVibrancyEffect_NotificationCenter {
...
+
+	[Export ("widgetList:didReorderRow:toRow:"), EventArgs ("NCWidgetListViewControllerDidReorder"), DefaultValue (false)]
+		void DidReorderRow (NCWidgetListViewController list, nuint row, nuint newIndex);
+
+	[Export ("widgetList:shouldRemoveRow:"), DelegateName ("NCWidgetListViewControllerShouldRemoveRow"), DefaultValue (false)]
+		bool ShouldRemoveRow (NCWidgetListViewController list, nuint row);
+
+	[Export ("widgetList:didRemoveRow:"), EventArgs ("NCWidgetListViewControllerDidRemoveRow"), DefaultValue (false)]
+		void DidRemoveRow (NCWidgetListViewController list, nuint row);
+	}
```
Comment (mandel-macaque) [inline src/notificationcenter.cs]: What happened with all the indentation? Maybe an editor issue?

File: tools/mmp/driver.cs [hunk 5]
```diff
@@ -484,6 +487,8 @@ static void Pack (IList<string> unprocessed)
 
 			if (registrar == RegistrarMode.Default)
 				registrar = RegistrarMode.Dynamic;
+			if (is_extension)
+				registrar = RegistrarMode.Static;
```
Comment (mandel-macaque) [inline tools/mmp/driver.cs]: I think we can do something like the following:

```csharp

registrar = is_extension ? RegistrarMode.Static : RegistrarMode.Dynamic;
```

But I don't know if we really need to do the registrar == Default or not. If it is needed just leave the code as it is.

File: tools/mmp/driver.cs [hunk 6]
```diff
@@ -555,7 +560,7 @@ static void Pack (IList<string> unprocessed)
 				GatherAssemblies ();
 				CheckReferences ();
 
-				if (!resolved_assemblies.Exists (f => Path.GetExtension (f).ToLower () == ".exe"))
+				if (!resolved_assemblies.Exists (f => Path.GetExtension (f).ToLower () == ".exe") && !is_extension)
 					throw new MonoMacException (79, true, "No executable was copied into the app bundle.  Please contact 'support@xamarin.com'", "");
```
Comment (mandel-macaque) [inline tools/mmp/driver.cs]: Would change the if to have !is_extension first to short cirtuit the && and not do the line expression.

File: msbuild/Xamarin.Mac.Tasks/Tasks/CompileAppManifest.cs [hunk 7]
```diff
@@ -38,12 +39,18 @@ public override bool Execute ()
 				return false;
 			}
 
+			if (AssemblyName.IndexOf ('.') != AssemblyName.LastIndexOf ('.'))
+				LogAppManifestError ("The assembly name ({0}) cannot contain more than one dot.", AssemblyName);
```
Comment (rolfbjarne) [inline msbuild/Xamarin.Mac.Tasks/Tasks/CompileAppManifest.cs]: What happens if an assembly has more than one dot?

I ran into something similar for iOS, and added the same check, but it turned out lots of people have assemblies with more than one dot, so it became problematic. After some investigation it turned out that the problem wasn't more than one dot, but executables that ended with `.app`.

File: msbuild/Xamarin.Mac.Tasks/Tasks/Mmp.cs (hunk 1 above)
Comment (rolfbjarne) [inline msbuild/Xamarin.Mac.Tasks/Tasks/Mmp.cs]: The opening brace also goes on the same line as the if.

File: msbuild/Xamarin.Mac.Tasks/Xamarin.Mac.Common.targets [hunk 8]
```diff
@@ -60,7 +65,7 @@ Copyright (C) 2014 Xamarin. All rights reserved.
 
 	<PropertyGroup>
 		<_CanOutputAppBundle>False</_CanOutputAppBundle>
-		<_CanOutputAppBundle Condition="'$(OutputType)' == 'Exe'">True</_CanOutputAppBundle>
+		<_CanOutputAppBundle Condition="'$(OutputType)' == 'Exe' Or '$(IsAppExtension)' == 'true'">True</_CanOutputAppBundle>
```
Comment (rolfbjarne) [inline msbuild/Xamarin.Mac.Tasks/Xamarin.Mac.Common.targets]: You're using `true` (lower-case) inside the condition and `True` elsewhere. IMHO the same casing should be used everywhere.

File: msbuild/Xamarin.Mac.Tasks/Xamarin.Mac.Common.targets [hunk 9]
```diff
@@ -70,6 +75,40 @@ Copyright (C) 2014 Xamarin. All rights reserved.
 	</PropertyGroup>
 
 	<PropertyGroup>
+		<ImplicitlyExpandDesignTimeFacades>true</ImplicitlyExpandDesignTimeFacades>
```
Comment (rolfbjarne) [inline msbuild/Xamarin.Mac.Tasks/Xamarin.Mac.Common.targets]: Same: true/True

File: msbuild/Xamarin.iOS.Tasks.Core/Xamarin.iOS.Common.targets [hunk 10]
```diff
@@ -1270,6 +1270,7 @@ Copyright (C) 2013-2016 Xamarin. All rights reserved.
 
 	<Target Name="_AssignAppExtensionConfiguration" Condition="'@(_AppExtensionReference)' != ''">
 		<!-- assign configs if building a solution file -->
+
```
Comment (rolfbjarne) [inline msbuild/Xamarin.iOS.Tasks.Core/Xamarin.iOS.Common.targets]: Unnecessary whitespace change.

File: msbuild/Xamarin.iOS.Tasks.Core/Xamarin.iOS.Common.targets [hunk 11]
```diff
@@ -1377,6 +1378,7 @@ Copyright (C) 2013-2016 Xamarin. All rights reserved.
 
 	<Target Name="_AssignWatchAppConfiguration" Condition="'@(_WatchAppReference)' != ''">
 		<!-- assign configs if building a solution file -->
+
```
Comment (rolfbjarne) [inline msbuild/Xamarin.iOS.Tasks.Core/Xamarin.iOS.Common.targets]: Whitespace.

File: runtime/Makefile [hunk 12]
```diff
@@ -356,6 +357,21 @@ $(eval $(call ObjTemplate,x86_64,64))
 $(eval $(call ObjTemplate,i386,DEBUG32,-debug,-DDEBUG))
 $(eval $(call ObjTemplate,x86_64,DEBUG64,-debug,-DDEBUG))
 
+$(foreach arch,$(MAC_ARCHITECTURES),.libs/mac/extension-main.$(arch).o): EXTRA_DEFINES=-DEXTENSION
+
+.libs/mac/extension-main.%.o:
+	echo .libs/mac/extension-main.%.o:
+	rm -f $@
+	clang -c -DEXTENSION extension-main.m -arch $* -o $@
```
Comment (rolfbjarne) [inline runtime/Makefile]: This is not quiet 😄

File: runtime/launcher.m [hunk 13]
```diff
@@ -499,7 +513,9 @@
 
 #define __XAMARIN_MAC_RELAUNCH_APP__ "__XAMARIN_MAC_RELAUNCH_APP__"
 
-int main (int argc, char **argv)
+extern "C" int NSExtensionMain (int argc, char** argv);
```
Comment (rolfbjarne) [inline runtime/launcher.m]: You can `#include "main-internal.h"` to avoid re-defininig this here.

File: runtime/launcher.m [hunk 14]
```diff
@@ -558,11 +574,26 @@ int main (int argc, char **argv)
...
 
 	return rv;
 }
+
+int main (int argc, char **argv)
+{
+	return xamarin_main (argc, argv, false);
+}
+
+extern "C" int extension_main (int argc, char **argv)
```
Comment (rolfbjarne) [inline runtime/launcher.m]: Public functions should be prefixed with `xamarin_` (also define the function in an internal header so that you can avoid the `extern "C"` part).

File: src/foundation.cs [hunk 15]
```diff
@@ -11368,15 +11368,20 @@ public interface NSPropertyListSerialization {
 		bool IsValidForFormat (NSObject plist, NSPropertyListFormat format);
 	}
 
-	[iOS (8,0)][Mac (10,10, onlyOn64 : true)] // Not defined in 32-bit
+#if XAMCORE_2_0 || !MONOMAC
```
Comment (rolfbjarne) [inline src/foundation.cs]: Isn't this a breaking change for XM/Classic?

File: src/notificationcenter.cs [hunk 16]
```diff
@@ -47,6 +64,107 @@ interface UIVibrancyEffect_NotificationCenter {
...
+
+	[Mac (10,10)]
+	[BaseType (typeof(NSViewController), Delegates=new string [] { "Delegate" }, Events=new Type [] { typeof (NCWidgetSearchViewDelegate)})]
+	interface NCWidgetSearchViewController
+	{
+		[Wrap ("WeakDelegate")]
+		[Protocolize]
```
Comment (rolfbjarne) [inline src/notificationcenter.cs]: Don't use `[Protocolize]` in new code, just bind using the protocol (`INCWidgetSearchViewDelegate`) instead.

File: src/notificationcenter.cs [hunk 17]
```diff
@@ -47,6 +64,107 @@ interface UIVibrancyEffect_NotificationCenter {
...
 	}
+#endif
+
+#if MONOMAC
+	[Mac (10,10)]
+	[BaseType (typeof(NSViewController), Delegates=new string [] { "Delegate" }, Events=new Type [] { typeof (NCWidgetListViewDelegate)})]
+	interface NCWidgetListViewController
+	{
+		[Wrap ("WeakDelegate")]
+		[Protocolize]
```
Comment (rolfbjarne) [inline src/notificationcenter.cs]: Don't use [Protocolize] in new code, just bind using the protocol (`INCWidgetListViewDelegate`) instead.

File: tools/mmp/mmp.csproj [hunk 18]
```diff
@@ -1,5 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
-<Project DefaultTargets="Build" ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+﻿<Project DefaultTargets="Build" ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Comment (rolfbjarne) [inline tools/mmp/mmp.csproj]: Did XS really remove the `<?xml` declaration?

File: tools/mmp/mmp.csproj [hunk 19]
```diff
@@ -314,6 +313,9 @@
     <Compile Include="..\..\src\ObjCRuntime\PlatformAvailability2.cs">
       <Link>external\PlatformAvailability2.cs</Link>
     </Compile>
+    <EmbeddedResource Include="config_mobile">
+       <LogicalName>config_mobile</LogicalName>
+    </EmbeddedResource>
```
Comment (rolfbjarne) [inline tools/mmp/mmp.csproj]: This looks like XS just moved one node to somewhere else in the file, and the changes to the entire file can be ignored.

File: runtime/launcher.m [hunk 20]
```diff
@@ -429,29 +430,42 @@
...
+        NSString *exePath;
+        if (plist != NULL)
+            exeName = (NSString *) [plist objectForKey:@"MonoBundleExecutable"];
+        else
+            fprintf (stderr, PRODUCT ": Could not find Info.plist in the bundle.\n");
+        
+        if (exeName == NULL)
+            exeName = [[NSString stringWithUTF8String: data->basename] stringByAppendingString: @".exe"];
+        
+        exePath = [[[NSString stringWithUTF8String: xamarin_get_bundle_path ()] stringByAppendingString: @"/"] stringByAppendingString: exeName];
```
Comment (rolfbjarne) [inline runtime/launcher.m]: You removed the `if (mkbundle) {` chunk, was that intentional?

File: msbuild/Xamarin.MacDev.Tasks.Core/Tasks/CodesignTaskBase.cs [hunk 21]
```diff
@@ -92,6 +98,7 @@ public override bool Execute ()
 			Log.LogTaskProperty ("ResourceRules", ResourceRules);
 			Log.LogTaskProperty ("SigningKey", SigningKey);
 			Log.LogTaskProperty ("ExtraArgs", ExtraArgs);
+			Log.LogTaskProperty("IsAppExtension", IsAppExtension);
```
Comment (VincentDondain) [inline msbuild/Xamarin.MacDev.Tasks.Core/Tasks/CodesignTaskBase.cs]: We're missing a space here before the parenthesis.

**Comment by xamarin-release-manager** [comment]
Build success

File: msbuild/Xamarin.Mac.Tasks/Xamarin.Mac.AppExtension.CSharp.targets (hunk 3 above)
Comment (timrisi) [inline msbuild/Xamarin.Mac.Tasks/Xamarin.Mac.AppExtension.CSharp.targets]: Currently there's no support to build mac projects in VS, only in XS on a mac

**Comment by xamarin-release-manager** [comment]
Build success

File: msbuild/Xamarin.Mac.Tasks/Tasks/CompileAppManifest.cs (hunk 7 above)
Comment (timrisi) [inline msbuild/Xamarin.Mac.Tasks/Tasks/CompileAppManifest.cs]: Looks like this was part of the code I copied over from iOS (when I first looked at extensions last may).  I'll update it to check for the .app instead like iOS is doing now

**Comment by xamarin-release-manager** [comment]
Build success

File: msbuild/Xamarin.Mac.Tasks/Xamarin.Mac.Common.targets (hunk 8 above)
Comment (timrisi) [inline msbuild/Xamarin.Mac.Tasks/Xamarin.Mac.Common.targets]: going with the lower case true over upper, since there's 30 instances of lc and only 5 uc

Comment (VincentDondain) [inline msbuild/Xamarin.Mac.Tasks/Xamarin.Mac.Common.targets]: Don't ask me why but it's looks like we do that everywhere in our iOS msbuild targets. Lower case inside condition and upper case as the property value.
So that's line isn't shocking to me.

File: runtime/launcher.m (hunk 20 above)
Comment (timrisi) [inline runtime/launcher.m]: Looks like it might have gotten wiped out into the rebase onto latest (at the time) master.  I'll get it fixed

File: tools/mmp/driver.cs (hunk 5 above)
Comment (timrisi) [inline tools/mmp/driver.cs]: With your suggested replacement if you had a build that wasn't an extension and had the registrar set to static or IL it'd get wiped out and set back to dynamic.

File: src/foundation.cs (hunk 15 above)
Comment (chamons) [inline src/foundation.cs]: Now fixed.

**Comment by xamarin-release-manager** [comment]
Build success
